### ISSUE TITLE
Dropdown does not conform to ARIA authoring practices

### ISSUE THREAD

**Issue by michaelwayneharris87** [twbs/bootstrap#28295]
The dropdown widget (as of 4.3.1) does not conform to the [ARIA 1.1 specification](https://www.w3.org/TR/wai-aria-1.1/#aria-haspopup). That should present some accessibility concerns.

Per the ARIA 1.1 spec, when a button with aria-haspopup is set to true, the popup itself **MUST** have a role=menu, and should conform to the [ARIA specification for the menu role](https://www.w3.org/TR/wai-aria-1.1/#menu). In general, the descendant options should have such roles as role=menuitem.

One solution would be simply adding the role=menu and role=menuitem to the dropdown and dropdown items. I'm hesitant to recommend that, however, as there is dropdown content such as headers, dividers, text, forms, etc. The ARIA spec isn't explicit, but I don't read it as allowing for the inclusion of that kind of content within an ARIA menu. None of the [menu button examples in the ARIA 1.1  authoring practices](https://www.w3.org/TR/wai-aria-practices-1.1/#menubutton) include such content, for example. 

Further, if you add role=menu and role=menuitem to the dropdown and dropdown items, you should take the menu items out of the tab focus order. The convention with ARIA widgets like this is that [composite UI components should have only one focusable element, and arrow keys move focus within the component](https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_general_within).

The other option may be to remove the aria-haspopup=true attribute altogether from the button altogether. In doing so, the dropdown stops following a menu button widget pattern and starts following a [disclosure widget pattern](https://www.w3.org/TR/wai-aria-practices-1.1/#disclosure). 

In contrast to the menu, if you go this route, you may want to remove support for arrows to navigate among items, as there's no convention to moving focus through the focusable items of a disclosure using arrow keys.

A third option is instead of implementing half of each ARIA widget into the dropdown component is to create more discreet widgets, each of which adheres to the ARIA patterns more closely.

I'm not sure what to recommend, since it might mean rethinking how the component would work. But whatever direction you choose, you should become more in line with the ARIA spec than you are now.

**Comment by patrickhlauke** [comment]
for the dropdowns, we've opted at this stage to treat them like disclosure widgets, not pop-ups https://www.w3.org/TR/wai-aria-practices-1.1/#disclosure, as they're often used for stuff that's not just a menu. menu has very limited range of what it can/can't contain: menu items, menu checkbox items, menu readio items, submenus, separators. as soon as an author wants to add something else (like structured content, or a form control (like a login form directly as part of the dropdown), the ARIA menu pattern won't allow it, which leads to unexpected non-standard results in browsers/AT. so, in order to keep it more generalised, bootstrap core does NOT implement the menu pattern per se. (see also https://patrickhlauke.github.io/aria/presentation/#150 slides 150-154 for a bit of high-level context)

**Comment by patrickhlauke** [comment]
and now that i've had my morning cup of coffee, i see that indeed our code still has `aria-haspopup` present, which is what the crux of the problem here is (sorry, i saw "isn't a menu" and went into explanation mode). we'll remove that (in future, we may move to treat dropdowns as dialogs, which will allow for mixed content, but will require further behaviours like focus management/trapping)

**Comment by patrickhlauke** [comment]
leaving the cursor key behaviour on the dropdowns - it's non-standard, but more as a "nice to have". unless it turns out to be more confusing for users than it's worth...
